jetty-io/src/main/java/org/eclipse/jetty/io/DefaultRetainableByteBufferPool.java

```diff
@@ -0,0 +1,390 @@
+        if (maxMemory > 0)
+        {
+            long excess = getMemory(direct) - maxMemory;
+            if (excess > 0)
+                evict(direct, excess);
+        }
+    }
+
+    /**
+     * This eviction mechanism searches for the RetainableByteBuffers that were released the longest time ago.
+     * @param direct true to search in the direct buffers buckets, false to search in the heap buffers buckets.
+     * @param excess the amount of bytes to evict. At least this much will be removed from the buckets.
+     */
+    private void evict(boolean direct, long excess)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("evicting {} bytes from {} pools", excess, (direct ? "direct" : "heap"));
+        long totalMemoryCleared = 0L;
+        long now = System.nanoTime();
+
+        Pool<RetainableByteBuffer>[] buckets = direct ? _direct : _indirect;
+        @SuppressWarnings("unchecked")
+        Pool<RetainableByteBuffer>.Entry[] oldestEntries = new Pool.Entry[buckets.length];
+
+        for (int i = 0; i < buckets.length; i++)
+        {
+            Pool<RetainableByteBuffer> bucket = buckets[i];
+            oldestEntries[i] = findOldestEntry(now, bucket);
+        }
+
+        while (true)
+        {
+            int oldestEntryBucketIndex = -1;
+            Pool<RetainableByteBuffer>.Entry oldestEntry = null;
+
+            for (int i = 0; i < oldestEntries.length; i++)
+            {
+                Pool<RetainableByteBuffer>.Entry entry = oldestEntries[i];
+                if (entry == null)
+                    continue;
+
+                if (oldestEntry != null)
+                {
+                    long entryAge = now - entry.getPooled().getLastUpdate();
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("checking entry of capacity {} and age {} vs entry of capacity {} and age {}", entry.getPooled().capacity(), now - entry.getPooled().getLastUpdate(), oldestEntry.getPooled().capacity(), now - oldestEntry.getPooled().getLastUpdate());
+                    if (entryAge > now - oldestEntry.getPooled().getLastUpdate())
+                    {
+                        oldestEntry = entry;
+                        oldestEntryBucketIndex = i;
+                    }
+                }
+                else
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("checking entry of capacity {} and age {}", entry.getPooled().capacity(), now - entry.getPooled().getLastUpdate());
+                    oldestEntry = entry;
+                    oldestEntryBucketIndex = i;
+                }
+            }
+
+            if (oldestEntry != null)
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("removing entry of capacity {} and age {}", oldestEntry.getPooled().capacity(), now - oldestEntry.getPooled().getLastUpdate());
+                oldestEntry.remove();
```

Comment:
What if `remove()` returns `false`? I don't think you can decrement the memory in that case.

Reply:
You're right, fixed that.